models/__init__.py

```diff
@@ -0,0 +1,121 @@
+import itertools
+import math
+
+import torch
+from torch.nn import functional as F
+
+from models.spectral_clustering import SpectralClusteringDiarizationModel
+from models.pyannote import PyannoteDiarizationModel
+
+# Code is adapted from pyannote.audio. All credit goes to pyannote.audio
+# https://github.com/pyannote/pyannote-audio/tree/master/pyannote/audio/models
+
+def rle1d(tensor):
+	#_notspeech_ = ~F.pad(speech, [1, 1])
+	#channel_i_channel_j = torch.cat([(speech & _notspeech_[..., :-2]).nonzero(), (speech & _notspeech_[..., 2:]).nonzero()], dim = -1)
+	#return [dict(begin = i / sample_rate, end = j / sample_rate, channel = channel) for channel, i, _, j in channel_i_channel_j.tolist()]
+
+	assert tensor.ndim == 1
+	starts = torch.cat(( torch.tensor([0], dtype = torch.long, device = tensor.device), (tensor[1:] != tensor[:-1]).nonzero(as_tuple = False).add_(1).squeeze(1), torch.tensor([tensor.shape[-1]], dtype = torch.long, device = tensor.device)))
+	starts, lengths, values = starts[:-1], (starts[1:] - starts[:-1]), tensor[starts[:-1]]
+	return starts, lengths, values
+
+
+def normalized_symmetric_laplacian(W):
+	# https://en.wikipedia.org/wiki/Laplacian_matrix#Symmetric_normalized_Laplacian
+	D_sqrt = W.sum(dim = -1).sqrt()
+	return torch.eye(W.shape[-1], device = W.device, dtype = W.dtype) - W / D_sqrt.unsqueeze(-1) / D_sqrt.unsqueeze(-2)
+
+
+def pdist(A, squared = False, eps = 1e-4):
+	normAsq = A.pow(2).sum(dim = -1, keepdim = True)
+	res = torch.addmm(normAsq.transpose(-2, -1), A, A.transpose(-2, -1), alpha = -2).add_(normAsq)
+	return res.clamp_(min = 0) if squared else res.clamp_(min = eps).sqrt_()
+
+
+def shrink(tensor, min):
+	return torch.where(tensor >= min, tensor, torch.zeros_like(tensor))
+
+
+def acf(K):
+	arange = torch.arange(len(K), device = K.device)
+	Z = (arange.unsqueeze(-1) - arange.unsqueeze(-2)).abs()
+	ACF = torch.zeros_like(arange, dtype = K.dtype).scatter_add_(0, Z.flatten(), K.flatten()).div_(Z.flatten().bincount(minlength = len(arange)))
+	M = ACF[Z.flatten()].view_as(Z)
+	return ACF, M
+
+
+def gaussian_kernel(kernel_size, sigma):
+	kernel = 1
+	meshgrids = torch.meshgrid([torch.arange(size, dtype=torch.float32) for size in kernel_size])
+	for size, std, mgrid in zip(kernel_size, sigma, meshgrids):
+		mean = (size - 1) / 2
+		kernel *= 1 / (std * math.sqrt(2 * math.pi)) * torch.exp(-((mgrid - mean) / (2 * std)) ** 2)
+	return (kernel / kernel.sum())[None, None, ...]
+
+
+def cdist(A, B, squared = False, eps = 1e-4):
+	normAsq = A.pow(2).sum(dim = -1, keepdim = True)
+	normBsq = B.pow(2).sum(dim = -1, keepdim = True)
+	res = torch.addmm(normBsq.transpose(-2, -1), A, B.transpose(-2, -1), alpha = -2).add_(normAsq)
+	return res.clamp_(min = 0) if squared else res.clamp_(min = eps).sqrt_()
+
+
+def cosine_kernel(E):
+	E = F.normalize(E, dim = 1)
+	return E @ E.t()
+
+
+def wang_affinity(E, gaussian_kernel_size = 9, gaussian_sigma = 1.0, row_threshold = 0.95, shrinking = 0.01):
```

Comment:
this method is duplicated in models/spectral_clustering.py. why?

Reply:
Because it is used in vis.ipynb. Outside of SpectralClusteringDiarizationModel class. Should be removed later, after vis.ipynb refactoring or removement.